CuPy: new version 13.5.1 and fixes

yizeyi18 opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
This PR adds CuPy released version 13.5.1, pre-released version 14.0.0a1, and synchonized several version constraint from its release pages. 
Some bugs occured during building CuPy itself or its dependencies, this PR also fixes them, including:
 - Migarate `install_option` to `configure_settings` for `py-fastrlock`
 - Constraint version of `py-jsonschema` for `rocprofiler-dev`
 - Disable mscclpp-related cmake options for `rccl`, see #724 
 - Add several rocm package include dirs for `py-cupy`

This PR makes `py-cupy` python-3.13 compatible, through adding new versions.

yizeyi18 committed:
py-cupy: update for python 3.13 compatibility

yizeyi18 committed:
add "main" version for py-cupy

yizeyi18 committed:
Merge branch 'spack:develop' into cupy-new

yizeyi18 committed:
add jsonschema version constraint, disable msccl, and add several include dirs

yizeyi18 commented:
Maintainers of related packages: 

  rocprofiler-dev & rccl:
  @afzpatel @renjithravindrankannath @srekolam

yizeyi18 commented:
@spackbot fix style

On my PC there are no style issue found. Why? Are there some extra cmdline flags needed to check the packages repo?
<img width="804" height="304" alt="image" src="https://github.com/user-attachments/assets/50a0f10a-6c7d-4590-a8a2-cbfb7b0dc941" />

yizeyi18 committed:
[@spackbot] updating style on behalf of yizeyi18

afzpatel reviewed (approved):
rccl and rocprofiler-dev changes LGTM. I'll try adding the mscclpp package and updating the rccl recipe accordingly.

tldahlgren reviewed (commented):
Branch and new sha256s have been confirmed.

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("py-numpy@1.22:2.0", when="@13.2", type=("build", "run"))
+    depends_on("py-numpy@1.22:2.2", when="@13.4", type=("build", "run"))
+    depends_on("py-numpy@1.22:2.3", when="@13.5", type=("build", "run"))
+    depends_on("py-numpy@1.24:2", when="@14", type=("build", "run"))
```
For `main` (https://github.com/cupy/cupy/blob/main/pyproject.toml), the upper bound is `2.5`, not `2` though I don't think that dependency is specified here.

yizeyi18 replied:
I specify only the major version `2` for two reasons:
 - The release notes for CuPy v14.0.0a1 only mentioned "CuPy v14’s behavior will be aligned with NumPy v2", with no explicitly specified version. As CuPy v14 had not been released officially, keep aligned with its (pre)release notes is not bad.
 - The newest release of `numpy` is `2.3.2`, and development of numpy 2.4.0 had just started for 2 months (see https://github.com/numpy/numpy/releases/tag/v2.4.0.dev0). I think "numpy < 2.6" here is a preemptive declaration, not meaning it's really bounded here or supported these (non-developed) versions. 
 
So for `cupy 14` the major version, `numpy @:2` is the only constraint that could be determined now.

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
-    depends_on("python@3.8:", when="@12:", type=("build", "run"))
+    depends_on("python@3.8:3.11", when="@12", type=("build", "run"))
+    depends_on("python@3.9:3.13", when="@13", type=("build", "run"))
+    depends_on("python@3.10:", when="@14", type=("build", "run"))
```
The `python` dependency range for `main` is `@3.9:`

yizeyi18 replied:
CuPy v14.0.0a1 release note says "Support for Python 3.9, NumPy 1.22 and 1.23, SciPy 1.7, 1.8, and 1.9 has been dropped."

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("py-scipy@1.6:1.11", when="@:12+all", type=("build", "run"))
+    depends_on("py-scipy@1.7:1.12", when="@13+all", type=("build", "run"))
+    depends_on("py-scipy@1.10:1.17", when="@14+all", type=("build", "run"))
+    depends_on("py-cython@0.29.22:0.29", when="@:13.3 +all", type=("build", "run"))
+    depends_on("py-cython@3:", when="@13.4: +all", type=("build", "run"))
```
The `py-cython` dependency also appears on lines 50-53, so please co-locate them.

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("py-numpy@1.24:2", when="@14", type=("build", "run"))
+    depends_on("py-scipy@1.6:1.11", when="@:12+all", type=("build", "run"))
+    depends_on("py-scipy@1.7:1.12", when="@13+all", type=("build", "run"))
+    depends_on("py-scipy@1.10:1.17", when="@14+all", type=("build", "run"))
```
The version range is inclusive so, from https://github.com/cupy/cupy/blob/v14.0.0a1/setup.py, this should be:

```suggestion
    depends_on("py-scipy@1.10:1.16", when="@14+all", type=("build", "run"))
```

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("py-numpy@1.22:2.3", when="@13.5", type=("build", "run"))
+    depends_on("py-numpy@1.24:2", when="@14", type=("build", "run"))
+    depends_on("py-scipy@1.6:1.11", when="@:12+all", type=("build", "run"))
+    depends_on("py-scipy@1.7:1.12", when="@13+all", type=("build", "run"))
```
From `pyproject.toml` for the version (https://github.com/cupy/cupy/blob/v13.5.1/pyproject.toml), this should be:

```suggestion
    depends_on("py-scipy@1.7:1.16", when="@13+all", type=("build", "run"))
```

Add a comment if that's not the case.

yizeyi18 replied:
You are right, I misadjusted here.

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("py-cython@0.29.22:0.29", when="@:13.3 +all", type=("build", "run"))
+    depends_on("py-cython@3:", when="@13.4: +all", type=("build", "run"))
     depends_on("py-optuna@2:", when="+all", type=("build", "run"))
+    depends_on("py-optuna@3:4", when="@13+all", type=("build", "run"))
```
According to `pyproject.toml` (https://github.com/cupy/cupy/blob/v13.5.1/pyproject.toml), this dependency isn't needed.

Where does this come from?

yizeyi18 replied:
Release note of CuPy v12 says support of optuna v2 is dropped. I cannot recall why I set `13` instead of `12`...
See: https://github.com/cupy/cupy/releases/tag/v12.0.0rc1

tldahlgren reviewed (changes requested):
There are some inconsistencies with the python dependencies (see comments). Add a comment in each case where there is a reason for the dependencies to be different from what's in the repository.

tldahlgren commented on repos/spack_repo/builtin/packages/py_cupy/package.py:
```diff
@@ -39,38 +43,52 @@ class PyCupy(PythonPackage, CudaPackage, ROCmPackage):
+    depends_on("python@3.8:3.11", when="@12", type=("build", "run"))
+    depends_on("python@3.9:3.13", when="@13", type=("build", "run"))
+    depends_on("python@3.10:", when="@14", type=("build", "run"))
     depends_on("py-setuptools", type="build")
```
This should be `py-setuptools@77:` for `main`. Have not checked for the other versions so at least each new version should be checked.

yizeyi18 replied:
I did not see constraint for version of `setuptools` before commit https://github.com/cupy/cupy/commit/4b711cb2308cbf6d242536b0bd59165844f3df8e . In release notes I found that `setuptools v74:` is supported since CuPy v13.4.0, but nothing else. 

I think now there could be an upper bound for `py-cupy@:13.3`, but wait for a determined lower bound from release notes(or official release).

yizeyi18 replied:
Or say, in other words, versions before `cupy@14` have no explicitly required `py-setuptools` version, and the requirement of `cupy@14` have not been concretized yet, so here leaves the constraint blank. 

The only pre-release `v14.0.0a1` also has no explicitly required `py-setuptools` version.

tldahlgren commented on repos/spack_repo/builtin/packages/py_fastrlock/package.py:
```diff
@@ -24,7 +25,19 @@ class PyFastrlock(PythonPackage):
     depends_on("py-setuptools", type="build")
     depends_on("py-cython", type="build")
```
There's a version range bound for `py-cython` for at least `v0.8.3` at https://github.com/scoder/fastrlock/blob/v0.8.3/requirements.txt.

```
Cython>=3.0.11, <3.1
setuptools
wheel
```

tldahlgren replied:
Another for `0.8.1` at https://github.com/scoder/fastrlock/blob/v0.8.1/requirements.txt.

tldahlgren replied:
Not seeing one in `setup.py` for `0.5` but am seeing it in https://github.com/scoder/fastrlock/blob/v0.5/requirements-appveyor.txt. (Didn't dig in to see if that's relevant.)

yizeyi18 replied:
Sorry for didn't take a look for version constraint here. You are right.

tldahlgren replied:
Np.

tldahlgren reviewed (commented):
Noted some additional issues related to `py-fastrlock`.

yizeyi18 committed:
Apply suggestions from code review

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

yizeyi18 committed:
apply suggestions from code review:
 - refine version constraint of py-setuptools for py-cupy
 - co-locate lines about py-cython in py-cupy
 - refine version constraint of py-cython in py-fastrlock

yizeyi18 committed:
resolve conflicts

yizeyi18 committed:
apply suggestions from code review
